src/packaging/metadata.py

```diff
@@ -406,3 +427,193 @@ def parse_email(data: Union[bytes, str]) -> Tuple[RawMetadata, Dict[str, List[st
+        message = email.message.EmailMessage()
+        message["content-type"] = value
+        content_type, parameters = (
+            message.get_content_type(),  # Defaults to `text/plain` if not parseable.
+            message["content-type"].params,
+        )
+        # Check if content-type is valid or defaulted to `text/plain` and thus was
+        # not parseable.
+        if content_type not in content_types or content_type not in value:
+            raise self._invalid_metadata(
+                f"{{field}} must be one of {list(content_types)}, not {value!r}"
+            )
+
+        charset = parameters.get("charset", "UTF-8")
+        if charset != "UTF-8":
+            raise self._invalid_metadata(
+                f"{{field}} can only specify the UTF-8 charset, not {list(charset)}"
+            )
+
+        markdown_variants = {"GFM", "CommonMark"}
+        variant = parameters.get("variant", "GFM")  # Use an acceptable default.
+        if content_type == "text/markdown" and variant not in markdown_variants:
+            raise self._invalid_metadata(
+                f"valid Markdown variants for {{field}} are {list(markdown_variants)}, "
+                f"not {variant!r}",
+            )
+        return value
+
+    def _process_dynamic(self, value: List[str]) -> List[str]:
+        for dynamic_field in map(str.lower, value):
+            if dynamic_field in {"name", "version", "metadata-version"}:
+                raise self._invalid_metadata(
+                    f"{value!r} is not allowed as a dynamic field"
+                )
+            elif dynamic_field not in _EMAIL_TO_RAW_MAPPING:
+                raise self._invalid_metadata(f"{value!r} is not a valid dynamic field")
+        return list(map(str.lower, value))
+
+    @staticmethod
+    def _process_provides_extra(
+        value: List[str],
+    ) -> List[utils.NormalizedName]:
+        return [utils.canonicalize_name(name, validate=True) for name in value]
+
+    _process_requires_python = staticmethod(specifiers.SpecifierSet)
+
+    @staticmethod
+    def _process_requires_dist(
+        value: List[str],
+    ) -> List[requirements.Requirement]:
+        return list(map(requirements.Requirement, value))
+
+
+class Metadata:
+    _raw: RawMetadata
+
+    @classmethod
+    def from_raw(cls, data: RawMetadata) -> "Metadata":
+        ins = cls()
+        ins._raw = data
+        return ins
+
+    @classmethod
+    def from_email(cls, data: Union[bytes, str]) -> "Metadata":
+        """Parse metadata from an email message."""
+        raw, unparsed = parse_email(data)
+        return cls.from_raw(raw)
+        # TODO Check `unparsed` for valid keys?
+
+    # TODO Check that fields are specified in a valid metadata version?
+    #      Should that be done per-field, or should it be done in a more
+    #      over-arching validate() method?
+
+    metadata_version: _Validator[_MetadataVersion] = _Validator()
+    name: _Validator[utils.NormalizedName] = _Validator()
```

Comment:
If I'm reading this correctly, I don't think we want to normalize the name here. The core metadata spec doesn't require normalization in general, just normalization prior to comparison, and I know from PyPI that people get very attached to the formatting of their name.

Should still validate that it's a valid value though.

Reply:
I take it you're thinking about having Warehouse get the name via this code and thus changing its format here would be bad? Otherwise my thinking was this avoided someone messing up the comparison later on by forgetting to normalize (i.e., normalize on the edges of your code before the data gets into your business logic).

Reply:
That's my plan yea, though I don't think it's *just* Warehouse because it can affect things like `pip show` and stuff too, but I suspect the people who care will primarily care about what Warehouse does, so it's *mostly* warehouse I guess.

At one point someone.. might have been you, talked about adding a `metadata.display_name()`, and that would honestly be fine with me too, though I have a slight preference for it just not canonicalizing because it just feels a little off to me? But that may just be my weirdness so if others feel differently I'm happy as long as I can still get the valid, but not canonicalized name.

Could always do something gross:

```python
class Name(str):
    def __eq__(self, other):
        return canoncalize_name(self) == canonicalize_name(other)
        
    # etc
```

Anyways, I'm cool with whatever you think here. I certainly wouldn't block it over that. That's just what I expect people will complain about.